### PR TITLE
Implement SegmentedCollectionsMarshal

### PR DESCRIPTION
**Pull request by sharwell**
Provides functionality from [`CollectionsMarshal`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal?view=net-8.0) and [`ImmutableCollectionsMarshal`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.immutablecollectionsmarshal?view=net-8.0) for custom data types defined in Roslyn.

### CODE REVIEW

File: src/Dependencies/Collections/ImmutableSegmentedDictionary`2+Builder+PrivateMarshal.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        /// Private helper class for use only by <see cref="SegmentedCollectionsMarshal"/>.
+        /// </summary>
+        internal static class PrivateMarshal
+        {
+            /// <inheritdoc cref="SegmentedCollectionsMarshal.GetValueRefOrNullRef{TKey, TValue}(ImmutableSegmentedDictionary{TKey, TValue}.Builder, TKey)"/>
+            public static ref TValue FindValue(Builder dictionary, TKey key)
+                => ref SegmentedCollectionsMarshal.GetValueRefOrNullRef(dictionary.GetOrCreateMutableDictionary(), key);
+
+            /// <inheritdoc cref="SegmentedCollectionsMarshal.GetValueRefOrAddDefault{TKey, TValue}(ImmutableSegmentedDictionary{TKey, TValue}.Builder, TKey, out bool)"/>
+            public static ref TValue? GetValueRefOrAddDefault(Builder dictionary, TKey key, out bool exists)
```
Comment (ToddGrun) [inline src/Dependencies/Collections/ImmutableSegmentedDictionary`2+Builder+PrivateMarshal.cs:21]: >GetValueRefOrAddDefault [](http://example.com/codeflow?start=38&length=23)

For the PR where I wanted this (72364), wouldn't GetValueRefOrNullRef have been sufficient as it just needs to be able to modify the colleciton while enumerating it?

Comment (ToddGrun) [reply src/Dependencies/Collections/ImmutableSegmentedDictionary`2+Builder+PrivateMarshal.cs:21]: nm, I need to read a bit more before commenting as I now see the FindValue method above for that purpose

File: src/Dependencies/Collections/ImmutableSegmentedDictionary`2+PrivateMarshal.cs [hunk 2]
```diff
@@ -40,6 +41,18 @@ internal static class PrivateInterlocked
 
                 return new ImmutableSegmentedDictionary<TKey, TValue>(dictionary);
             }
+
+            /// <inheritdoc cref="SegmentedCollectionsMarshal.GetValueRefOrNullRef{TKey, TValue}(ImmutableSegmentedDictionary{TKey, TValue}, TKey)"/>
+            public static ref readonly TValue FindValue(ImmutableSegmentedDictionary<TKey, TValue> dictionary, TKey key)
+                => ref SegmentedCollectionsMarshal.GetValueRefOrNullRef(dictionary._dictionary, key);
```
Comment (ToddGrun) [inline src/Dependencies/Collections/ImmutableSegmentedDictionary`2+PrivateMarshal.cs:47]: >=> ref SegmentedCollectionsMarshal.GetValueRefOrNullRef(dictionary._dictionary, key); [](http://example.com/codeflow?start=16&length=85)

I assume you like this better than having SegmentedCollectionMarshal use AsSegmentedDictionary directly?

File: src/Dependencies/Collections/SegmentedArray`1.cs [hunk 3]
```diff
@@ -429,8 +429,6 @@ public TestAccessor(SegmentedArray<T> array)
             }
 
             public static int SegmentSize => SegmentedArray<T>.SegmentSize;
-
-            public T[][] Items => _array._items;
```
Comment (ToddGrun) [inline src/Dependencies/Collections/SegmentedArray`1.cs:433]: >Items [](http://example.com/codeflow?start=25&length=5)

Does this class still get constructed anywhere?

File: src/Dependencies/Collections/SegmentedDictionary`2+PrivateMarshal.cs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        /// <inheritdoc cref="SegmentedCollectionsMarshal.GetValueRefOrNullRef{TKey, TValue}(SegmentedDictionary{TKey, TValue}, TKey)"/>
+        public static ref TValue FindValue(SegmentedDictionary<TKey, TValue> dictionary, TKey key)
+            => ref dictionary.FindValue(key);
+
+        /// <inheritdoc cref="SegmentedCollectionsMarshal.GetValueRefOrAddDefault{TKey, TValue}(SegmentedDictionary{TKey, TValue}, TKey, out bool)"/>
+        /// <remarks>
+        /// ⚠️ This method is mirrored by <see cref="SegmentedDictionary{TKey, TValue}.TryInsert(TKey, TValue, InsertionBehavior)"/>.
+        /// If you make any changes here, make sure to keep that version in sync as well.
+        /// </remarks>
+        public static ref TValue? GetValueRefOrAddDefault(SegmentedDictionary<TKey, TValue> dictionary, TKey key, out bool exists)
```
Comment (ToddGrun) [inline src/Dependencies/Collections/SegmentedDictionary`2+PrivateMarshal.cs:27]: >GetValueRefOrAddDefault [](http://example.com/codeflow?start=34&length=23)

If I don't need this right now, do you know of other code that does? It's a fair amount of code if we don't have a known use for it.

File: src/Dependencies/Collections/ImmutableSegmentedDictionary`2+PrivateMarshal.cs (hunk 2 above)
Comment (sharwell) [reply src/Dependencies/Collections/ImmutableSegmentedDictionary`2+PrivateMarshal.cs:47]: Didn't think of that. I'll test it and see.

File: src/Dependencies/Collections/SegmentedArray`1.cs (hunk 3 above)
Comment (sharwell) [reply src/Dependencies/Collections/SegmentedArray`1.cs:433]: It does not, so I'll simplify further.

File: src/Dependencies/Collections/ImmutableSegmentedDictionary`2+PrivateMarshal.cs (hunk 2 above)
Comment (sharwell) [reply src/Dependencies/Collections/ImmutableSegmentedDictionary`2+PrivateMarshal.cs:47]: ➡️ Will leave this one. I'd remove it if this was the only method in the `PrivateMarshal` class, but there are many others.

File: src/Dependencies/Collections/SegmentedDictionary`2+PrivateMarshal.cs (hunk 4 above)
Comment (sharwell) [reply src/Dependencies/Collections/SegmentedDictionary`2+PrivateMarshal.cs:27]: ➡️ Removing from this branch, and will create a different branch to add it if we need it in the future

File: src/Dependencies/Collections/ImmutableSegmentedDictionary`2+PrivateMarshal.cs [hunk 5]
```diff
@@ -40,6 +41,18 @@ internal static class PrivateInterlocked
...
+
+            /// <inheritdoc cref="SegmentedCollectionsMarshal.AsImmutableSegmentedDictionary{TKey, TValue}(SegmentedDictionary{TKey, TValue}?)"/>
+            internal static ImmutableSegmentedDictionary<TKey, TValue> AsImmutableSegmentedDictionary(SegmentedDictionary<TKey, TValue>? dictionary)
+                => dictionary is not null ? new ImmutableSegmentedDictionary<TKey, TValue>(dictionary) : default;
+
+            /// <inheritdoc cref="SegmentedCollectionsMarshal.AsSegmentedDictionary{TKey, TValue}(ImmutableSegmentedDictionary{TKey, TValue})"/>
+            internal static SegmentedDictionary<TKey, TValue>? AsSegmentedDictionary(ImmutableSegmentedDictionary<TKey, TValue> dictionary)
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Dependencies/Collections/ImmutableSegmentedDictionary`2+PrivateMarshal.cs:54]: ➡️ Yes, this matches the behavior we see for similar types in the BCL.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for reviews

**Comment by sharwell** [comment]
Tests are added.

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for reviews

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for reviews

**Comment by ToddGrun** [comment]
Gentle nudge to @dotnet/roslyn-compiler (I've got a decently impactful allocation improvement which requires this to be done first)

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for reviews

**Review by 333fred — APPROVED** [review]

**Review by jjonescz — APPROVED** [review]

**Pull request merged**